prototypes/pipemarker.lua

```diff
@@ -0,0 +1,149 @@
+    current_entity.name = 'pipelayer-pipe-dot' .. directions
+    if direction_index == '0' then
+        current_entity.final_render_layer = 'light-effect'
+        --? The single pipe dot is twice as large so must be scaled
+        current_entity.animation.scale = 0.5
+    else
+      current_entity.animation.width = 32
+      current_entity.animation.height = 32
+    end
+    current_entity.animation.filename = '__pipelayer__/graphics/entity/markers/pipe-marker-dot' .. directions .. '.png'
+    new_dots[#new_dots + 1] = current_entity
+end
+
+for _, stuff in pairs(new_dots) do
+  data:extend {
+      merge {
+          base_entity,
+          stuff
+      }
+  }
+end
+
+local underground_marker_beams = {}
+local marker_beams = util.table.deepcopy(data.raw['beam']['electric-beam-no-sound'])
+marker_beams.name = 'pipelayer-pipe-marker-beam'
+marker_beams.width = 1.0
+
+--? Ensure beam doesn't fade when in use
+marker_beams.damage_interval = 2000000000
+--? Don't need the beam to conduct an action
+marker_beams.action = nil
+
+--? Start and ending properties must be present but have to match framecount of body and other properties
+marker_beams.start = {
+    filename = '__core__/graphics/empty.png',
+    line_length = 1,
+    width = 1,
+    height = 1,
+    frame_count = 1,
+    axially_symmetrical = false,
+    direction_count = 1,
+    hr_version = {
+        filename = '__core__/graphics/empty.png',
+        line_length = 1,
+        width = 1,
+        height = 1,
+        frame_count = 1,
+        axially_symmetrical = false,
+        direction_count = 1
+    }
+}
+marker_beams.ending = {
+    filename = '__core__/graphics/empty.png',
+    line_length = 1,
+    width = 1,
+    height = 1,
+    frame_count = 1,
+    axially_symmetrical = false,
+    direction_count = 1,
+    hr_version = {
+        filename = '__core__/graphics/empty.png',
+        line_length = 1,
+        width = 1,
+        height = 1,
+        frame_count = 1,
+        axially_symmetrical = false,
+        direction_count = 1
+    }
+}
+
+--? Head tail and body must all have same frame count
+marker_beams.head = {
+    filename = '__pipelayer__/graphics/entity/markers/pipe-marker-horizontal.png',
+    line_length = 1,
+    width = 64,
+    height = 64,
+    frame_count = 1,
+    animation_speed = 1,
+    scale = 0.5
+}
+marker_beams.tail = {
+    filename = '__pipelayer__/graphics/entity/markers/pipe-marker-horizontal.png',
+    line_length = 1,
+    width = 64,
+    height = 64,
+    frame_count = 1,
+    animation_speed = 1,
+    scale = 0.5
+}
+--? Body is an array of sprite definitions instead of a single sprite definition
+marker_beams.body = {
+    {
+        filename = '__pipelayer__/graphics/entity/markers/pipe-marker-horizontal.png',
+        line_length = 1,
+        width = 64,
+        height = 64,
+        frame_count = 1,
+        scale = 0.5
+    }
+}
+underground_marker_beams[#underground_marker_beams + 1] = marker_beams
```

Comment:
I assume this is a vestige of PipeTools supporting multiple colors, since as written here underground_marker_beams only has one element?

Reply:
Yep. Lots of copy pasta. Could just wrap the data extend in a table inside the parens and dump the table index all together.